pymarkdown/plugins/rule_md_049.py

```diff
@@ -0,0 +1,278 @@
+"""
+import os.path
+import re
+import urllib.parse
+from typing import Callable, Dict, Generator, List, Optional, Pattern, Tuple, cast
+
+from pymarkdown.inline_markdown_token import ReferenceMarkdownToken, TextMarkdownToken
+from pymarkdown.leaf_markdown_token import SetextHeadingMarkdownToken
+from pymarkdown.markdown_token import MarkdownToken
+from pymarkdown.plugin_manager.plugin_details import PluginDetails
+from pymarkdown.plugin_manager.plugin_scan_context import PluginScanContext
+from pymarkdown.plugin_manager.rule_plugin import RulePlugin
+
+NOT_A_VALID_HEADING = "Wrong reference: # anchor not a valid heading"
+
+FILE_DOES_NOT_EXIST = "Dangling reference: file does not exist!"
+
+TriggeredRuleParams = Tuple[str, int, int, str, str, str, Optional[str]]
+
+
+class RuleMd049(RulePlugin):
+    """
+    Class to implement a plugin that looks for link and images elements that have an invalid uri.
+    """
+
+    EXTERNAL_LINK_RE = re.compile("^(.+):.*$")
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.headline_processor = HeadlineProcessor()
+        self.anchor_validator = AnchorValidator(
+            self.report_next_token_error, self._create_triggered_rule_params
+        )
+
+    def completed_file(self, context: PluginScanContext) -> None:
+        """
+        Event that the file being currently scanned is now completed.
+        """
+        scan_file = os.path.realpath(context.scan_file)
+
+        self.headline_processor.completed_file(scan_file)
+        current_file_headlines = self.headline_processor.get_headlines(scan_file)
+        triggers = self.anchor_validator.complete_file(
+            scan_file, current_file_headlines
+        )
+        for trigger_params in triggers:
+            context.add_triggered_rule(*trigger_params)
+
+    def get_details(self) -> PluginDetails:
+        """
+        Get the details for the plugin.
+        """
+        return PluginDetails(
+            plugin_name="validate-refs",
+            plugin_id="MD049",
+            plugin_enabled_by_default=False,
+            plugin_description="Local URIs should be valid",
+            plugin_version="0.0.2",
+            plugin_interface_version=1,
+            plugin_url="https://github.com/jackdewinter/pymarkdown/blob/main/docs/rules/rule_md049.md",
+        )
+
+    def next_token(self, context: PluginScanContext, token: MarkdownToken) -> None:
+        """
+        Event that a new token is being processed.
+        """
+        self.headline_processor.next_token(token)
+        self._validate_references(context, token)
+
+    def _validate_references(
+        self, context: PluginScanContext, token: MarkdownToken
+    ) -> None:
+        if not token.is_inline_image and not token.is_inline_link:
+            # Not a reference so nothing to do here
+            return
+
+        link_uri, link_anchor = self._extract_link_uri(token)
+        if self.EXTERNAL_LINK_RE.match(link_uri):
```

Comment:
Maybe rename to either `EXTERNAL_LINK_REGEX` or just `EXTERNAL_LINK`. Personally, I find `RE` to be confusing